File: lib/domain/cultivation_plan/policies/task_schedule_item_update_policy.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Policies
      module TaskScheduleItemUpdatePolicy
        module_function

        # @param attributes_seed [Hash] 文字列キー
        # @param amount_snapshot [Domain::CultivationPlan::Dtos::TaskScheduleItemAmountSnapshot]
        # @param calculator [Domain::CultivationPlan::Calculators::AmountUnitConversionCalculator]
        # @param rescheduled_at [Time] 日付変更時に Gateway へ渡すタイムスタンプ（edge で注入）
        # @return [Hash] Gateway update 用（文字列キー）
        def build_update_attributes(attributes_seed:, amount_snapshot:, calculator:, rescheduled_at:)
          attributes = attributes_seed.transform_keys(&:to_s)
          if attributes.key?("scheduled_date") && attributes["scheduled_date"].present?
            new_date = Domain::CultivationPlan::Policies::TaskScheduleItemCreatePolicy.parse_scheduled_date!(
              attributes["scheduled_date"]
            )
            attributes["scheduled_date"] = new_date

            if amount_snapshot.scheduled_date != new_date
              attributes["rescheduled_at"] = rescheduled_at
              attributes["status"] = Domain::AgriculturalTask::Constants::TaskScheduleItemStatuses::RESCHEDULED
            end
          end

          converted = calculator.apply_to_update_attributes(
            attributes: attributes,
            current_amount: amount_snapshot.amount,
            current_unit: amount_snapshot.amount_unit,
            new_unit: attributes["amount_unit"],
            amount_param: attributes["amount"]
          )
          converted || attributes
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/policies/task_schedule_item_update_policy.rs
//! Ruby: `Domain::CultivationPlan::Policies::TaskScheduleItemUpdatePolicy`

use std::collections::BTreeMap;

use time::{Date, OffsetDateTime};

use crate::agricultural_task::constants::task_schedule_item_statuses::RESCHEDULED;
use crate::cultivation_plan::calculators::amount_unit_conversion_calculator::AmountUnitConversionCalculator;
use crate::cultivation_plan::dtos::TaskScheduleItemAmountSnapshot;
use crate::cultivation_plan::policies::task_schedule_item_create_policy::parse_scheduled_date;

fn format_date(date: Date) -> String {
    format!(
        "{:04}-{:02}-{:02}",
        date.year(),
        u8::from(date.month()),
        date.day()
    )
}

fn format_datetime(dt: OffsetDateTime) -> String {
    dt.unix_timestamp().to_string()
}

/// Ruby: `TaskScheduleItemUpdatePolicy.build_update_attributes`
pub fn build_update_attributes(
    attributes_seed: &BTreeMap<String, String>,
    amount_snapshot: &TaskScheduleItemAmountSnapshot,
    calculator: &AmountUnitConversionCalculator,
    rescheduled_at: OffsetDateTime,
) -> BTreeMap<String, String> {
    let mut attributes = attributes_seed.clone();

    if let Some(raw) = attributes.get("scheduled_date").filter(|s| !s.trim().is_empty()) {
        if let Ok(new_date) = parse_scheduled_date(raw) {
            attributes.insert("scheduled_date".into(), format_date(new_date));
            if amount_snapshot.scheduled_date != new_date {
                attributes.insert("rescheduled_at".into(), format_datetime(rescheduled_at));
                attributes.insert("status".into(), RESCHEDULED.to_string());
            }
        }
    }

    let converted = calculator.apply_to_update_attributes(
        &attributes,
        amount_snapshot.amount,
        amount_snapshot.amount_unit.as_deref(),
        attributes.get("amount_unit").map(String::as_str),
        attributes.get("amount").map(String::as_str),
    );
    converted.unwrap_or(attributes)
}

#[cfg(test)]
mod policies_task_schedule_item_update_policy_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/cultivation_plan/policies_task_schedule_item_update_policy_test.rs"));
}
